tests/test_types_typeddict.py

```diff
@@ -378,3 +378,32 @@ class Model(BaseModel):
             Model(d=input_value)
     else:
         assert Model(d=input_value).d == expected
+
+
+class RecursiveTypedDict(typing_extensions.TypedDict):
+    foo: Optional['RecursiveTypedDict']
+
+
+class RecursiveTypedDictModel(BaseModel):
+    rec: RecursiveTypedDict
+
+
+def test_recursive_typeddict():
+    assert RecursiveTypedDictModel(rec={'foo': {'foo': None}}).rec == {'foo': {'foo': None}}
+
+
+T = TypeVar('T')
+
+
+class RecursiveGenTypedDict(typing_extensions.TypedDict, Generic[T]):
+    foo: Optional['RecursiveGenTypedDict[T]']
+    ls: List[T]
+
+
+class RecursiveGenTypedDictModel(BaseModel, Generic[T]):
+    rec: RecursiveGenTypedDict[T]
+
+
+def test_recursive_generic_typeddict():
+    data: RecursiveGenTypedDict[int] = {'foo': {'foo': None, 'ls': []}, 'ls': []}
+    assert RecursiveGenTypedDictModel[int](rec=data).rec == data
```

Comment:
can you expand this test to show what happens if the "inner" _and_ "outer" values of `ls` are equal to `[1]`, and `['a']`? If it isn't validating properly, you can make it an xfail or add a TODO comment, and I'll get it working when I work on generic dataclasses.

Reply:
Can you also add a TODO: comment indicating to make sure it works if `RecursiveGenTypedDict` was a forward reference? Don't worry about even writing the test, I'll do that